### PR TITLE
Add component runtime selector

### PR DESCRIPTION
**Pull request by rottencandy**
## Fixes 
https://issues.redhat.com/browse/HAC-1605

## Description
Add a runtime selector to components review section. That uses the devfile sample API to fetch list of runtimes.
- Allows selecting a runtime when CDQ detection fails to detect one automatically.
- Allows optionally selecting a different runtime if the automatic detection is not satisfactory.

## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Scenario 1: Automatic detection failed

https://user-images.githubusercontent.com/20013884/194301862-e7d82b9a-322c-42bd-9bed-66efb0c39dd2.mp4

### Scenario 2: Detection succeeded

https://user-images.githubusercontent.com/20013884/194301877-931bcb0b-c21c-4aa2-aac4-1adae2c22801.mp4

## How to test or reproduce?
Use a non-detectable repo, or try to change the runtime after detection has happened.

## Browser conformance: 

- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by rottencandy** [comment]
cc @openshift/team-hacdev-ux

**Comment by itsptk** [comment]
@rottencandy In the second video 'Scenario 2: Detection succeeded' I guess I was expecting the dropdown to be pre-populated with the detected runtime. Is that not what's happening here? Does that ever happen?

**Comment by beaumorley** [comment]
@rottencandy In the first video 'Scenario 1: Automatic detection failed' would it be possible to put the error text "Runtime not detected" in the red error message area and not within the form field. And put the "Select a runtime" that is now in red within the field? So swap them?

**Comment by rottencandy** [comment]
> @rottencandy In the second video 'Scenario 2: Detection succeeded' I guess I was expecting the dropdown to be pre-populated with the detected runtime. Is that not what's happening here? Does that ever happen?

Fixed. The pre population might not always happen since we're currently fetching runtimes from the devfile samples API, and it doesn't provide all the available runtimes. But I think it should work fine when we switch to the devfile stacks API in the future.

**Comment by rottencandy** [comment]
> @rottencandy In the first video 'Scenario 1: Automatic detection failed' would it be possible to put the error text "Runtime not detected" in the red error message area and not within the form field. And put the "Select a runtime" that is now in red within the field? So swap them?

Updated.
s

File: src/components/ImportForm/ReviewSection/RuntimeSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  const onChange = (value: string) => {
+    const runtime = samples.find((s) => s.name === value);
+    setSelectedRuntime(runtime);
+    setRuntimeSource((runtime.attributes?.git as any)?.remotes?.origin);
+    setDetecting(true);
+    setFieldValue('isDetected', false);
+  };
+
+  const [detectedComponents, detectionLoaded, detectionError] = useComponentDetection(
```
[the remark this answers is not in the document]
Comment (rottencandy) [reply src/components/ImportForm/ReviewSection/RuntimeSelector.tsx:57]: Fixed

File: src/components/ImportForm/ReviewSection/ReviewSection.tsx [hunk 2]
```diff
@@ -140,6 +128,8 @@ const ReviewSection: React.FunctionComponent = () => {
             key={component.componentStub.componentName}
             detectedComponent={component}
             detectedComponentIndex={index}
+            isExpanded={!isDetected}
+            showRuntimeSelector
```
[the remark this answers is not in the document]
Comment (rottencandy) [reply src/components/ImportForm/ReviewSection/ReviewSection.tsx:132]: I used this prop to hide the runtime selector in the component settings page, should we be showing it there as well?

[the remark this answers is not in the document]
Comment (rohitkrai03) [reply src/components/ImportForm/ReviewSection/ReviewSection.tsx:132]: Okay, got it. No we shouldn't show runtime selector in settings page.

**Comment by rohitkrai03** [comment]
@rottencandy The PR checks are failing. Seems like an e2e test failure. Can you take a look?

**Review by rohitkrai03 — APPROVED** [review]
/approve

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Okay, I found a few issues while testing multi component detection.

- All the tiles are expanded by default.
- Some of the runtime selectors say `Select runtime` even though it was successfully detected.

https://user-images.githubusercontent.com/6041994/194536258-11e4804c-8461-4df5-a2fd-b1e34bcf846f.mov

**Comment by rottencandy** [comment]
Fixed. Dropdown will now show `Other` when projectType does not match any of the samples.

https://user-images.githubusercontent.com/20013884/194560703-0e969b6c-ab81-4ea3-8fc5-a3e391b61cb8.mp4

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Updating any one of the runtime selectors in a multi component detection it changes the selection for all and shows error message.

https://user-images.githubusercontent.com/6041994/194576440-80589a51-da31-466c-b452-e867cf45bdec.mov

**Comment by rottencandy** [comment]
> Updating any one of the runtime selectors in a multi component detection it changes the selection for all and shows error message.

Fixed.

**Review by rohitkrai03 — APPROVED** [review]
/approve

**Pull request merged**